src/Interpreters/HashJoin.cpp

```diff
@@ -964,7 +978,18 @@ void HashJoin::joinBlockImpl(
     constexpr bool need_filter = !need_replication && (inner || right || (is_semi_join && left) || (is_anti_join && left));
 
     /// Rare case, when keys are constant or low cardinality. To avoid code bloat, simply materialize them.
-    Columns materialized_keys = JoinCommon::materializeColumns(block, key_names_left);
+    Columns materialized_keys;
+    if (!cast_keys_info.empty())
+    {
+        /// Also need to convert left columns
```

Comment:
You're converting left columns, not right? Converting the right one would probably be more efficient -- IIRC, the right table is hashed and is expected to be small compared to the left table, so it makes sense to convert the lesser one.

Reply:
Ok, for HashJoin it's changed to casting columns from right table